sisl/geom/nanoribbon.py

```diff
@@ -154,3 +159,461 @@ def zgnr(width, bond=1.42, atoms=None):
     agnr : armchair graphene nanoribbon
     """
     return graphene_nanoribbon(width, bond, atoms, kind='zigzag')
+
+
+Heteroribbon_section = namedtuple("Section", ["W", "L", "shift", "atoms"], defaults=(None, 1, 0, None))
+
+
+@set_module("sisl.geom")
+def heteroribbon(sections, bond, atoms=None, kind="armchair"):
+    """Builds a nanoribbon with multiple widths.
+
+    The function makes sure that only valid ribbons, that is
+    ribbons with no dangling bonds, are being produced.
+    Explanatory errors are raised if that's not possible. Only the 
+    connection at the periodic boundary remains unchecked.
+
+    It only works for armchair ribbons for now.
+
+    Parameters
+    ----------
+    sections: array-like of (Heteroribbon_section or array-like)
+        A list of sections to be added to the ribbon.
+
+        Each section is either a `Heteroribbon_section` or an array that will
+        be converted to a `Heteroribbon_section`. Therefore it should contain:
+            W: int
+                The width of the section.
+            L: int, optional
+                The number of units of the section. Note that a "unit" is
+                not a unit cell, but half of it. I.e. a zigzag string.
```

Comment:
This should be remembered to be changed for the other direction, since then it is a bit different.